### PR TITLE
#168781707 create and view accommodation

### PR DESCRIPTION
**Pull request by alainmateso**
[Finishes#168781707]

#### What does this PR do?
Give the user the ability to add new accommodation facilities and view the facilities they have created
#### Description of Task to be completed?
- adds endpoint for creating new accommodation
- adds endpoint for viewing user's accommodations
- provide tests for the feature

#### How should this be manually tested?
- Clone the repository from [here](https://github.com/andela/caret-bn-backend.git)
- Browse to the repo directory
- navigate to ft-create-accommodation-168781707 branch
- Run `npm install`
- Refer to .env.example and add `CLOUDINARY_URL=`
- Go to the following url https://developers.google.com/oauthplayground/ and authorize the following APIs (This is done in step 1 of the Playground process).
 a) Google OAuth apiv2
      - https://www.googleapis.com/auth/userinfo.email
      - https://www.googleapis.com/auth/userinfo.profile
- Exchange the authorization code for tokens (This is done in step 2 of the Playground process)
- Copy the new access token for your `GOOGLE_ACCESS_TOKEN=` in your .env file

- run `sequelize db:migrate`
- run `sequelize db:seed:all`
- run `npm run dev`
 **To add a new accommodation**
- In Postman Provide a login bearer token
- Fill the body with all necessary fields and select image file(s)
- send a `POST` request to `http://localhost:<port>/api/v1/accommodation`
**To view a user's accommodations**
- In Postman Provide a login bearer token
- send a `GET` request to `http://localhost:<port>/api/v1/accommodation/myAccommodations`

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#168781707
](https://www.pivotaltracker.com/story/show/168781707)
#### Screenshots (if appropriate)
Screen Shot 2019-10-16 at 11 46 47
Screen Shot 2019-10-16 at 11 47 16
Screen Shot 2019-10-16 at 11 47 29

#### Questions:
N/A

### CODE REVIEW

**Comment by Cheza-Dzabala** [comment]
Please add instructions in your PR about getting an ` api_key `.

**Comment by Cheza-Dzabala** [comment]
Hey Alain, 
So I realized I'm able to create the same accomodation over and over over. Can you find a way of restricting that. Maybe two businesses cannot exist at the same address? Because as is, I kept on hitting send and the business duplicates.

**Review by Cheza-Dzabala — CHANGES_REQUESTED** [review]
Hey Alain, 
So I realized I'm able to create the same accomodation over and over over. Can you find a way of restricting that. Maybe two businesses cannot exist at the same address? Because as is, I kept on hitting send and the business duplicates.

**Comment by Cheza-Dzabala** [comment]
All is well now. Cheers.

**Review by Cheza-Dzabala — APPROVED** [review]

**Review by ngireric123 — COMMENTED** [review]
Alain you did a good job but did u remember to do swagger in this API

**Review by nignanthomas — CHANGES_REQUESTED** [review]
Good job @alainmateso 
However I think **availableSpace** and **location** should be integers because you might later on want to make **calculations** on **availableSpace**. And owner displaying **owner details** instead of just a number.
Don't you think?

**Review by sabin18 — APPROVED** [review]
@alainmateso It looks good .

File: src/controllers/accommodationController.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+const { uploadImages, uploadSingle } = imagesHelper;
+
+export default class AccommodationController {
+  static async createAccommodation(req, res) {
+    const ownerId = req.user.payload.id;
+    let images = [];
+    const {
+      name, description, location, availableSpace, cost, highlights, amenities
+    } = req.body;
```
Comment (nakiwuge) [inline src/controllers/accommodationController.js:17]: Add a new line here

File: src/controllers/accommodationController.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    if (req.files.image.length > 1) { images = await uploadImages(req.files.image); } else {
+      images = await uploadSingle(req, res);
+      images = images.url;
+    }
```
Comment (nakiwuge) [inline src/controllers/accommodationController.js:22]: Add a new line here

File: src/controllers/accommodationController.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      name: name.toLowerCase(),
+      description,
+      location: location.toLowerCase(),
+      availableSpace,
+      cost,
+      highlights,
+      amenities,
+      owner: ownerId,
+      images
+    };
```
Comment (nakiwuge) [inline src/controllers/accommodationController.js:34]: Add a new line here

File: src/controllers/accommodationController.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    try {
+      const newAccommodation = await models.accommodations.create(accommodation);
+      return responseUtil(res, 201, strings.accommodation.success.CREATED, newAccommodation);
+    } catch (error) { return responseError(res, 400, error); }
+  }
+
+  static async getMyAccommodations(req, res) {
+    const owner = req.user.payload.id;
+    const MyAccommodations = await models.accommodations.findAll({ where: { owner } });
```
Comment (nakiwuge) [inline src/controllers/accommodationController.js:46]: Add a new line here

File: src/middlewares/inputValidation.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  static validateImage(req, res, next) {
+    if (!req.files.image) {
+      return responseError(res, 400, strings.images.NO_IMAGE);
+    }
+    return next();
+  }
+
+  static validateLogin(req, res, next) {
```
Comment (nakiwuge) [inline src/middlewares/inputValidation.js:43]: I thought you already have a  function for login validation

File: src/middlewares/inputValidation.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    });
+    validation(req, res, schema, next);
+  }
+
+  static async validateExistence(req, res, next) {
+    let { name, location } = req.body;
+    name = name.toLowerCase();
+    location = location.toLowerCase();
+    const foundAccommodation = await models.accommodations.findOne({ where: { name, location } });
+    if (foundAccommodation) {
```
Comment (nakiwuge) [inline src/middlewares/inputValidation.js:56]: Add a new line

File: src/routes/api/accommodations.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+const router = express.Router();
+const multipartMiddleware = multipart();
+
+const { createAccommodation, getMyAccommodations } = accommodationController;
+const { validateAddNew, validateImage, validateExistence } = InputValidation;
+
+router.post('/', verifyToken, multipartMiddleware, validateAddNew, validateExistence, validateImage, createAccommodation);
+router.get('/myAccommodations', verifyToken, getMyAccommodations);
+
+
```
Comment (nakiwuge) [inline src/routes/api/accommodations.js:18]: Remove this line

**Review by nakiwuge — CHANGES_REQUESTED** [review]
@alainmateso good job on the implementation whoever Kindly work on the feedback below.
- Add a role field to the user model. This will later be linked to the user modal.
- Ensure that only a supplier can create  accommodation
- Ensure that admins can view all accommodation facilities

I have also added other comments on your PR

File: src/middlewares/inputValidation.js (hunk 5 above)
Comment (alainmateso) [reply src/middlewares/inputValidation.js:43]: I have removed it and relocated inside that class in practice of putting all input validation into one class.

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**